### PR TITLE
feat: adding method for full ERC20 withdraw

### PR DESCRIPTION
**Pull request by Nilay27**
* Adding method for the complete withdrawal of ERC20 tokens from the inherited smart contracts for `WithdrawERC20.sol`

### CODE REVIEW

File: apps/smart-contracts/token/test/TokenShop.test.ts [hunk 1]
```diff
@@ -738,7 +738,7 @@ describe('TokenShop', () => {
   })
   // TODO : add tests for tx.origin vs msg.sender
 
-  describe('# withdrawERC20', () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/token/test/TokenShop.test.ts:741]: lets move this to the previous PR.

File: apps/smart-contracts/token/test/Vesting.test.ts [hunk 2]
```diff
@@ -1005,7 +1005,7 @@ describe('Vesting', () => {
     })
   })
 
-  describe('# withdrawERC20', () => {
+  describe('# withdrawERC20Amounts', () => {
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/token/test/Vesting.test.ts:1008]: Same here, move to previous PR.

File: packages/prepo-shared-contracts/contracts/WithdrawERC20.sol [hunk 3]
```diff
@@ -25,4 +25,23 @@ contract WithdrawERC20 is IWithdrawERC20, SafeOwnable, ReentrancyGuard {
       }
     }
   }
+
+  function withdrawERC20(address[] calldata _erc20Tokens)
```
Comment (ramenforbreakfast) [inline packages/prepo-shared-contracts/contracts/WithdrawERC20.sol:29]:
```suggestion
  function withdrawERC20Balance(address[] calldata _erc20Tokens)
```

Should be consistent and name this function as well to differentiate between them.

**Review by ramenforbreakfast — APPROVED** [review]
LGTM besides nits, should hold off on Xavier's review before merging

**Comment by ramenforbreakfast** [comment]
Rebased off updated PR #421 since i forgot to change the testcase section names.

File: apps/smart-contracts/token/test/TokenShop.test.ts [hunk 4]
```diff
@@ -782,6 +782,45 @@ describe('TokenShop', () => {
...
+      const externalERC20Recipient = user1.address
+      const externalERC20Decimals = 18
+      const mockERC20InitialSupply = parseEther(`100`)
+      externalERC20Token = await mockERC20Fixture(
+        'External ERC20',
+        'ExtERC20',
+        externalERC20Decimals,
+        externalERC20Recipient,
+        mockERC20InitialSupply
+      )
```
Comment (XavierEkkel) [inline apps/smart-contracts/token/test/TokenShop.test.ts:798]: why not just use payment token instead of setting up a new token?

File: packages/prepo-shared-contracts/test/WithdrawERC20.test.ts [hunk 5]
```diff
@@ -168,4 +167,66 @@ describe('WithdrawERC20', () => {
       )
     })
   })
+
+  describe('# withdrawERC20Balance', async () => {
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/test/WithdrawERC20.test.ts:171]: dont forget to update all the 'describe' names

File: packages/prepo-shared-contracts/contracts/interfaces/IWithdrawERC20.sol [hunk 6]
```diff
@@ -18,4 +18,11 @@ interface IWithdrawERC20 {
     address[] calldata erc20Tokens,
     uint256[] calldata amounts
   ) external;
+
+  /**
+   * @notice Withdraws entire balance of each ERC20 in `erc20Tokens`.
+   * @dev Only callable by `owner()`.
+   * @param erc20Tokens ERC20 tokens to be withdrawn
+   */
+  function withdrawERC20Balance(address[] calldata erc20Tokens) external;
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/contracts/interfaces/IWithdrawERC20.sol:27]:
```suggestion
  function withdrawERC20(address[] calldata erc20Tokens) external;
```

File: packages/prepo-shared-contracts/test/WithdrawERC20.test.ts [hunk 7]
```diff
@@ -16,6 +16,7 @@ describe('WithdrawERC20', () => {
   let withdrawERC20: WithdrawERC20
   let firstMockERC20: MockContract<Contract>
   let secondMockERC20: MockContract<Contract>
+  let ERC20ContractArray: string[]
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/test/WithdrawERC20.test.ts:19]: nit:
```suggestion
  let erc20ContractArray: string[]
```

File: packages/prepo-shared-contracts/test/WithdrawERC20.test.ts [hunk 8]
```diff
@@ -168,4 +167,66 @@ describe('WithdrawERC20', () => {
...
+
+    it('reverts if not owner', async () => {
+      expect(await withdrawERC20.owner()).to.not.eq(user1.address)
+
+      await expect(
+        withdrawERC20.connect(user1).withdrawERC20Balance(ERC20ContractArray)
+      ).revertedWith('Ownable: caller is not the owner')
+    })
+
+    it('succeeds if token balance = 0', async () => {
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/test/WithdrawERC20.test.ts:185]: also add the '= 0' case for multiple tokens?

File: packages/prepo-shared-contracts/test/WithdrawERC20.test.ts [hunk 9]
```diff
@@ -168,4 +167,66 @@ describe('WithdrawERC20', () => {
...
     })
   })
+
+  describe('# withdrawERC20Balance', async () => {
+    beforeEach(async () => {
+      await setupWithdrawERC20AndMockERC20()
+      ERC20ContractArray = [firstMockERC20.address, secondMockERC20.address]
+    })
+
+    it('reverts if not owner', async () => {
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/test/WithdrawERC20.test.ts:177]: also test an empty array? (if this is even possible to pass as an argument)

File: packages/prepo-shared-contracts/test/WithdrawERC20.test.ts [hunk 10]
```diff
@@ -168,4 +167,66 @@ describe('WithdrawERC20', () => {
...
+
+      await withdrawERC20.connect(owner).withdrawERC20Balance([firstMockERC20.address])
+
+      expect(await firstMockERC20.balanceOf(withdrawERC20.address)).to.eq(0)
+      expect(await firstMockERC20.balanceOf(owner.address)).to.eq(
+        ownerFirstERC20BalanceBefore.add(contractFirstERC20BalanceBefore)
+      )
+    })
+
+    it('withdraws entire balance of multiple ERC20 tokens', async () => {
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/test/WithdrawERC20.test.ts:208]: also test the same token twice

**Review by XavierEkkel — APPROVED** [review]
LGTM after comments
